File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/services/BookingResultCreditingTest.java
package it.giannibombelli.wsc2026.giftcard.application.services;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardCrediting;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.policies.CreditGiftCardPolicy;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;

import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.createBooking;
import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.getSavedGiftCard;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class BookingResultCreditingTest {

    private GiftCardRepository repository;
    private BookingResultCrediting creditGiftCardService;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);

        CreditGiftCardPolicy policy = new CreditGiftCardPolicy();
        GiftCardCrediting useCase = new GiftCardCrediting(repository);
        creditGiftCardService = new BookingResultCrediting(policy, useCase);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullParameters() {
            CreditGiftCardPolicy policy = new CreditGiftCardPolicy();
            GiftCardCrediting useCase = new GiftCardCrediting(repository);

            assertThatThrownBy(() -> new BookingResultCrediting(null, useCase))
                .isInstanceOf(NullPointerException.class);

            assertThatThrownBy(() -> new BookingResultCrediting(policy, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class BookingResultsHandling {
        @Test
        void onConfirmed_shouldIncreaseBalance() {
            GiftCard giftCard = getSavedGiftCard(repository);
            Booking booking = createBooking();
            Money credit = new Money(new BigDecimal("14.00"));

            creditGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingConfirmed(booking.id(), giftCard.id(), credit)
            );

            Optional<GiftCard> updated = repository.findById(giftCard.id());
            assertThat(updated).isPresent();
            assertThat(updated.get().balance()).isEqualTo(credit);
        }

        @Test
        void onRefused_shouldIncreaseBalance() {
            GiftCard giftCard = getSavedGiftCard(repository);
            final Booking booking = createBooking();
            Money credit = new Money(new BigDecimal("9.99"));

            creditGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingRefused(booking.id(), giftCard.id(), credit)
            );

            GiftCard after = repository.findById(giftCard.id()).orElseThrow();
            assertThat(after.balance()).isEqualTo(credit);
        }

        @Test
        void onRejected_shouldDoNothing() {
            GiftCard giftCard = getSavedGiftCard(repository);
            Booking booking = createBooking();
            Money amount = new Money(new BigDecimal("5.00"));

            creditGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingRejected(booking.id(), giftCard.id(), amount)
            );

            GiftCard persisted = repository.findById(giftCard.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(Money.zero());
        }

        @Test
        void shouldFailIfCardDoesNotExist() {
            GiftCardId nonExisting = EntityId.generate(GiftCardId::new);
            Booking booking = createBooking();
            Money amount = new Money(new BigDecimal("5.00"));

            assertThatThrownBy(() -> creditGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingConfirmed(booking.id(), nonExisting, amount)
            ))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}

File: typescript/tests/giftcard/application/services/bookingResultCrediting.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import {
  bookingConfirmed,
  bookingRefused,
  bookingRejected,
} from '@/booking/domain/events/bookingResultEvents.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { BookingResultCrediting } from '@/giftcard/application/services/bookingResultCrediting.js';
import { GiftCardCrediting } from '@/giftcard/application/usecases/giftCardCrediting.js';
import { CreditGiftCardPolicy } from '@/giftcard/application/policies/creditGiftCardPolicy.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';
import { GiftCardAggregateFactory } from '../../../testsupport/giftcard/aggregateFactory.js';

describe('BookingResultCrediting', () => {
  let repository: SqliteGiftCardRepository;
  let service: BookingResultCrediting;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    const policy = new CreditGiftCardPolicy();
    const useCase = new GiftCardCrediting(repository);
    service = new BookingResultCrediting(policy, useCase);
  });

  function createBooking(): BookingId {
    return generateId((value) => new BookingId(value));
  }

  describe('construction', () => {
    it('rejects null parameters', () => {
      const policy = new CreditGiftCardPolicy();
      const useCase = new GiftCardCrediting(repository);

      expect(() => new BookingResultCrediting(null as unknown as CreditGiftCardPolicy, useCase)).toThrow();
      expect(() => new BookingResultCrediting(policy, null as unknown as GiftCardCrediting)).toThrow();
    });
  });

  describe('booking results handling', () => {
    it('on confirmed should increase balance', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const bookingId = createBooking();
      const credit = new Money(14);

      service.handleBookingResults(bookingConfirmed(bookingId, giftCard.id().value.value, credit));

      const updated = repository.findById(giftCard.id());
      expect(updated).not.toBeNull();
      expect(updated!.balance()).toEqual(credit);
    });

    it('on refused should increase balance', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const bookingId = createBooking();
      const credit = new Money(9.99);

      service.handleBookingResults(bookingRefused(bookingId, giftCard.id().value.value, credit));

      const after = repository.findById(giftCard.id());
      expect(after).not.toBeNull();
      expect(after!.balance()).toEqual(credit);
    });

    it('on rejected should do nothing', () => {
      const giftCard = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const bookingId = createBooking();
      const amount = new Money(5);

      service.handleBookingResults(bookingRejected(bookingId, giftCard.id().value.value, amount));

      const persisted = repository.findById(giftCard.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(Money.zero());
    });

    it('should fail if card does not exist', () => {
      const nonExisting = generateId((value) => new GiftCardId(value));
      const bookingId = createBooking();
      const amount = new Money(5);

      expect(() =>
        service.handleBookingResults(bookingConfirmed(bookingId, nonExisting.value.value, amount))
      ).toThrow();
    });
  });
});
